### ISSUE TITLE
Mix silently ignores invalid options

### ISSUE THREAD

**Issue by xxiii23** [elixir-lang/elixir#2547]
Instead of generating an error, mix silently ignores/discards invalid options (at least with mix new):

mix new test_mod -module TEST_mod (works, but should have been --module, and module gets named TestMod))
mix new test_mod --bleagidyblah TEST_mod (works, same as above)

**Comment by nurugger07** [comment]
@josevalim is anyone currently working this one? I assume not but wanted to verify before I started looking into it

**Comment by josevalim** [comment]
No! The first step is to improve the option parser! :)

**Comment by nurugger07** [comment]
@josevalim I wouldn't mind taking a look at that. Has there been any discussion on expected behavior? I didn't see any open issues for OptionParser

**Comment by josevalim** [comment]
No, there is nothing yet. Working on a POC and proposing a solution is part of the task! :D

**Comment by nurugger07** [comment]
@josevalim cool, well I'm willing to help if needed. Just let me know

**Comment by josevalim** [comment]
Awesome @nurugger07! If you would like to explore the OptionParser API and write a proposal on how you think it should be improved, we would love to read it. I think @alco may have some ideas too.

**Comment by felipesere** [comment]
Has there been any progress on this?
Would a proposal simply be a PR?

**Comment by jisaacstone** [comment]
I've always liked pythons [ArgumentParser](https://docs.python.org/2/library/argparse.html). [Here is a quick POC](https://github.com/jisaacstone/ex_argument_parser) I wrote up to show how I think that might work in elixir.

Ended up being very different from OptionParser, but I'd love to hear you're thoughts.

Here are a few I have:
 - print help & exit is default. Makes sense in Python but Elixir maybe not so much. better to raise or return error?
- `strict` is an option in the struct but perhaps having a `parse!` method is a more idiomatic solution
- naming things is hard. should it be ArgumentParser or ArgParse?
